### PR TITLE
Migrate users from shell -> profile settings

### LINKED ISSUE #123171: Migrate users off shell, shellArgs and automationShell terminal settings
See https://github.com/microsoft/vscode/issues/123159

### PR DESCRIPTION
**Pull request by meganrogge**
This PR fixes #123171

### CODE REVIEW

**Comment by meganrogge** [comment]
To do: use shellArgs instead of automationShell once it's been brought back

File: src/vs/platform/terminal/common/terminal.ts [hunk 1]
```diff
@@ -21,9 +22,11 @@ export const enum TerminalSettingId {
 	ShellArgsLinux = 'terminal.integrated.shellArgs.linux',
 	ShellArgsMacOs = 'terminal.integrated.shellArgs.osx',
 	ShellArgsWindows = 'terminal.integrated.shellArgs.windows',
+	ShellArgs = 'terminal.integrated.shellArgs.',
```
Comment (meganrogge) [inline src/vs/platform/terminal/common/terminal.ts:25]: pull out into a diff't const enum

File: src/vs/workbench/contrib/terminal/browser/terminalService.ts [hunk 2]
```diff
@@ -68,6 +73,7 @@ export class TerminalService implements ITerminalService {
 	private _remoteTerminalsInitPromise: Promise<void> | undefined;
 	private _localTerminalsInitPromise: Promise<void> | undefined;
 	private _connectionState: TerminalConnectionState;
+	private _platform: string | undefined;
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminalService.ts:76]: remove

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 3]
```diff
@@ -1986,33 +2027,33 @@ registerThemingParticipant((theme: IColorTheme, collector: ICssStyleCollector) =
...
-		`);
+				.monaco - workbench.pane - body.integrated - terminal.xterm.xterm - viewport:: -webkit - scrollbar - thumb: hover { background - color: ${scrollbarSliderHoverBackgroundColor}; }
+			.monaco - workbench.pane - body.integrated - terminal.xterm - viewport: hover { scrollbar - color: ${scrollbarSliderHoverBackgroundColor} transparent; }
+			`);
 	}
 
 	const scrollbarSliderActiveBackgroundColor = theme.getColor(scrollbarSliderActiveBackground);
 	if (scrollbarSliderActiveBackgroundColor) {
-		collector.addRule(`.monaco-workbench .pane-body.integrated-terminal .xterm .xterm-viewport::-webkit-scrollbar-thumb:active { background-color: ${scrollbarSliderActiveBackgroundColor}; }`);
+		collector.addRule(`.monaco - workbench.pane - body.integrated - terminal.xterm.xterm - viewport:: -webkit - scrollbar - thumb: active { background - color: ${scrollbarSliderActiveBackgroundColor}; } `);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts]: This thing again 😱, has happened to me a couple of times too

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 4]
```diff
@@ -353,6 +356,44 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
+			!!this._configurationService.getValue(TerminalSettingPrefix.DefaultProfile + platform);
+		if (shouldMigrateToProfile && this._storageService.getBoolean(SHOULD_PROMPT_FOR_PROFILE_MIGRATION_KEY, StorageScope.WORKSPACE, true)) {
+			this._notificationService.prompt(
+				Severity.Info,
+				nls.localize('terminalProfileMigration', "You are using deprecated shell/shellArgs settings, but have a defaultProfile set. Want to create a profile from those and use that instead?"),
+				[
+					{
+						label: nls.localize('switchToProfile', "Yes"),
+						run: async () => {
+							const shell = this._configurationService.getValue(`${TerminalSettingPrefix.Shell + platform}`);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:372]: Can remove the string here

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 5]
```diff
@@ -353,6 +356,44 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
+							const shellArgs = this._configurationService.getValue(`${TerminalSettingPrefix.ShellArgs + platform}`);
+							const profile = await this._terminalProfileResolverService.createProfileFromShellAndShellArgs(shell, shellArgs);
+							if (profile) {
+								this._configurationService.updateValue(`${TerminalSettingPrefix.DefaultProfile + platform}`, profile.profileName);
+								this._configurationService.updateValue(`${TerminalSettingPrefix.Shell + platform}`, null);
+								this._configurationService.updateValue(`${TerminalSettingPrefix.ShellArgs + platform}`, null);
+								this._logService.trace(`migrated from shell/shellArgs, ${shell} ${shellArgs} to profile ${JSON.stringify(profile)}`);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:379]: shellArgs probably won't get logged right

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 6]
```diff
@@ -353,6 +356,44 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
 
+	async showProfileMigrationNotification(): Promise<void> {
+		const platform = this._getPlatformKey();
+		const shouldMigrateToProfile = (!!this._configurationService.getValue(TerminalSettingPrefix.Shell + platform) ||
+			!!this._configurationService.getValue(TerminalSettingPrefix.ShellArgs + platform)) &&
+			!!this._configurationService.getValue(TerminalSettingPrefix.DefaultProfile + platform);
+		if (shouldMigrateToProfile && this._storageService.getBoolean(SHOULD_PROMPT_FOR_PROFILE_MIGRATION_KEY, StorageScope.WORKSPACE, true)) {
+			this._notificationService.prompt(
+				Severity.Info,
+				nls.localize('terminalProfileMigration', "You are using deprecated shell/shellArgs settings, but have a defaultProfile set. Want to create a profile from those and use that instead?"),
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:367]: How about this?

This terminal is using the deprecated shell/shellArgs settings, do you want to migrate it to a profile?

[Migrate] [Don't show again]

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts]: I disabled prettier and had fixed that... I don't know how it happened again 😦

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 5 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:379]: they're just a string, array, or null. will that not work? when I tested, it seemed  fine

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts (hunk 3 above)
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts]: Ah, this is the cause 65e1707d19d4df5b1500ae260ceae9fad6d0cd96. I'm going to disable as I'm not a fan of format on save

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/terminalInstance.ts]: I think you might be looking at an old version? I fixed that earlier I think.

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 7]
```diff
@@ -1408,7 +1449,7 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 			});
 			this._storageService.store(SUGGESTED_RENDERER_TYPE, 'auto', StorageScope.GLOBAL, StorageTarget.MACHINE);
 		} catch (e) {
-			this._logService.warn(`Webgl could not be loaded. Falling back to the canvas renderer type.`, e);
+			this._logService.warn(`Webgl could not be loaded.Falling back to the canvas renderer type.`, e);
```
Comment (jsonify) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts]: This used to have a space at "...loaded.Falling back..." but its no longer there and is now incorrect spacing.

File: src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts [hunk 8]
```diff
@@ -410,6 +410,49 @@ export abstract class BaseTerminalProfileResolverService implements ITerminalPro
...
 	}
+
+	async createProfileFromShellAndShellArgs(shell?: unknown, shellArgs?: unknown): Promise<ITerminalProfile | undefined> {
+		const detectedProfile = this._terminalService.availableProfiles?.find(p => p.path === shell);
+		const fallbackProfile = (await this.getDefaultProfile({
+			remoteAuthority: this._remoteAgentService.getConnection()?.remoteAuthority,
+			os: this._primaryBackendOs!
+		}));
+		const profile = detectedProfile || fallbackProfile;
+		const args = typeof shellArgs === 'string' || Array.isArray(shellArgs) ? shellArgs : profile.args;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts:421]: Can use `_isValidShellArgs` here which also makes sure string is only used on Windows.

File: src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts [hunk 9]
```diff
@@ -410,6 +410,49 @@ export abstract class BaseTerminalProfileResolverService implements ITerminalPro
...
+		const createdProfile = {
+			profileName: profile.profileName,
+			path: profile.path,
+			args,
+			isDefault: true
+		};
+		return this._profilesMatch(createdProfile, detectedProfile) ? undefined : createdProfile;
+	}
+
+	private _profilesMatch(createdProfile: ITerminalProfile, detectedProfile?: ITerminalProfile): boolean {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts:431]: Might be better to inline this as it doesn't check if the profiles are the same, but only the path and args

**Review by Tyriar — APPROVED** [review]

**Pull request merged**